docs: add tutorial for reading plans

wjones127 opened the pull request:
The goal of the tutorial is to get someone to the point where they can read a serialized Substrait plan. SQL is used as a comparison since I assume most readers will have basic familiarity. JSON is used as the output format since it's used in some consumers as the input format for test.

wjones127 commented:
Not sure if this addresses exactly what #215 is getting at. But as a newcomer, I found it hard to figure out where to start if I wanted to read plans. I started drafting this tutorial (still in very rough form) to try that out.

Does the JSON output seem correct and the right way to go? Do we feel like the serialized protobuf bytes should also be provided?

I based the JSON output on this snippet:

```python
import ibis
from ibis_substrait.compiler.core import SubstraitCompiler
from google.protobuf.json_format import MessageToJson

compiler = SubstraitCompiler()

t = ibis.table([("b", "string"), ("a", "int64")], name="my_database.my_table")
result = compiler.compile(t.filter(t.a > 10))
print(MessageToJson(result))

result_bytes = result.SerializeToString()
```

westonpace commented on site/docs/tutorial/sql_to_substrait.md:
```diff
@@ -0,0 +1,245 @@
+for newcomers.
+
+Substrait can be a type, an expression, a relation, or a plan. We'll look at them in
+those order, going from simplest to most complex.
```
```suggestion
that order, going from simplest to most complex.
```

westonpace commented on site/docs/tutorial/sql_to_substrait.md:
```diff
@@ -0,0 +1,245 @@
+queries. We'll present the Substrait in JSON form so it is (somewhat) readable
+for newcomers.
+
+Substrait can be a type, an expression, a relation, or a plan. We'll look at them in
```
Substrait is a spec for plans.   A plan is made up of those building blocks but I'm not sure it is valid to say "Substrait can be a type, an expression, a relation, or a plan".

westonpace commented on site/docs/tutorial/sql_to_substrait.md:
```diff
@@ -0,0 +1,245 @@
+  "selection": {
+    "directReference": {
+      "structField": {
+        "field": 1
```
Might be worth noting that Substrait refers to fields numerically based on some kind of schema / field order (depends on the operator).  In SQL field references are always named references.

wjones127 replied:
Good point. I'll make sure to call that out. I'll need to look a little deeper to, since the natural question is "relative to *which* schema?"

westonpace commented on site/docs/tutorial/sql_to_substrait.md:
````diff
@@ -0,0 +1,245 @@
+```json
+{
+  "literal": {
+    "i8": 10,
````
Why is it an `i8`?

wjones127 replied:
:shrug: that's what Ibis emitted and it seemed fine to me. Though I suppose usually when people represent an integer it's an i32 or i64, so maybe it would be less weird if it was either of those.

I should probably explain this type to the reader (perhaps that's what you are getting at). One question I have: To what extent should a Substrait producer strive to keep the operands of the same type? If `a` in `a > 10` is an `i64` is it okay if I label `10` as a `i32`?

westonpace commented on site/docs/tutorial/sql_to_substrait.md:
```diff
@@ -0,0 +1,245 @@
+  "extensionFunction": {
+    "extensionUriReference": 1,
+    "functionAnchor": 2,
+    "name": "gt"
```
It might be a good idea to explicitly mention here that `gt` is equivalent to `>` in your above SQL expression.  Not everyone will be hard-wired to see "gt" and think "greater than".

wjones127 replied:
Good point. I'll also point out where the YAML definition is.

westonpace commented on site/docs/tutorial/sql_to_substrait.md:
```diff
@@ -0,0 +1,245 @@
+
+## Relations
+
+An logical or physical plan is typically represented as a tree of nodes, such
```
```suggestion
A logical or physical plan is typically represented as a tree of nodes, such
```
Do we want to introduce or mention "logical" and "physical"?  Won't these terms confuse a reader at this point?

wjones127 replied:
It's not clear here, but I was speaking of plans in general, not Substrait plans specifically. I assume that anyone coming from the SQL world has seen some form of logical / physical plan (at the very least, has run `EXPLAIN ...` at some point and tried to read the output). I'll make this more specific so that's clear.

westonpace commented on site/docs/tutorial/sql_to_substrait.md:
```diff
@@ -0,0 +1,245 @@
+
+An logical or physical plan is typically represented as a tree of nodes, such
+as filter, project, scan, or join. Each of these nodes is a **Relation** in 
+Substrait.
```
I feel like, as a naive reader, I might be confused here.  What is a relation exactly?  A part of a tree?  Why is it a tree?  What does this tree represent?  How is a relation different from a function?

westonpace commented on site/docs/tutorial/sql_to_substrait.md:
````diff
@@ -0,0 +1,245 @@
+```
+
+Read relations provide a schema in `baseSchema` as discussed in the prior section.
+The type of read is a named table, so the `namedTable` field is present with 
````
The type of read?  There is more than one? <-- Something a naive reader might be thinking here.

westonpace commented on site/docs/tutorial/sql_to_substrait.md:
```diff
@@ -0,0 +1,245 @@
+The type of read is a named table, so the `namedTable` field is present with 
+`names` containing the list of name segments (`my_database.my_table`).
+
+Read relations are leaf nodes, so other relations will have them nested within.
+For example, a filter relation of (`a > 10`) will look like:
```
Maybe...

Read relations are leaf nodes.  Leaf nodes don't depend on any other node for data and usually represent a source of data in our plan.  Leaf nodes are then typically used as input for other nodes that manipulate the data.

waynexia replied:
Do you think we need to refer to the docs https://substrait.io/relations/logical_relations/ where has a table shows the input and output number of each relation? Like read relation here require 0 input so that's why it's a leaf relation.

westonpace commented on site/docs/tutorial/sql_to_substrait.md:
```diff
@@ -0,0 +1,245 @@
+
+## Plans
+
+First, let's look at reading a table:
```
I feel like the first paragraph should explain what a plan is and not be giving an example.

westonpace commented on site/docs/tutorial/sql_to_substrait.md:
````diff
@@ -0,0 +1,245 @@
+SELECT * FROM my_table;
+```
+
+A plan has a list of relations, and each top one is a root relation. The root
+relation just has one input relation and then the names of the fields.
````
These are facts, and they are true, but I'm not sure what the story you are trying to tell is.  Perhaps something like...

Many plans are represented by a single tree of relations.  The root of the tree is the output of the plan and the leaves in the tree are the sources of data.

Howevr, some plans actually contain multiple trees.  One reason for this is when a plan has a common subtree that occurs multiple times throughout the plan.  That subtree can be pulled out and described on its own and then can be referenced in other trees.  This is why a Substrait plan is actually a list of relations.  Only one of these is a "root relation" and it forms the root of the plan.

westonpace reviewed (commented):
Thanks!  I think this will be quite helpful and is a good start.  I left a few comments on the content itself.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/substrait-io/substrait?pullRequest=354) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/substrait-io/substrait?pullRequest=354) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/substrait-io/substrait?pullRequest=354) it.</sub>

waynexia commented:
Super thanks! It's a lifesaver 🥰 As a beginner to substrait I have a lot less doubts after reading this article. Can't wait to see it land!

Z-Shang commented on site/docs/tutorial/expression_trees.svg:
```diff
@@ -0,0 +1,115 @@
+      <rect width="160" height="30" x="0" y="0" class="node" fill="#eee" />
+      <rect width="80" height="30" x="0" y="0" class="node" fill="steelblue" />
+      <text class="node light" x="40" y="20">Field Ref</text>
+      <text class="node" x="120" y="20">quantity</text>
```
shouldn't this field name to be discount?

wjones127 commented on site/docs/tutorial/sql_to_substrait.md:
```diff
@@ -0,0 +1,1305 @@
+at the end to select a subset of columns. However, the project relation only
+adds columns; it is not responsible for subsetting or reordering columns.
+
+<!-- TODO: Is this true? -->
+Instead, any relation can reorder or subset columns through the `emit` property.
+By default, it is set to `direct`, which outputs all columns "as is". But it can
+also be specified as a sequence of field indices.
```
Just want to make sure? Is this actually true? Is **any** relation able to subset or reorder columns?

westonpace replied:
Yes.  This is quite different from some existing tooling in this area (e.g. Acero, dplyr, ibis) and so if you are familiar with those then it can be a bit odd.  It has been a consistent source of confusion :).

It is a very "logical" concept and allows for writing compact plans.  One way to think about it is to recognize that just about any relation might be the terminal point for a particular column:

 * You might load a column purely for the sake of applying a filter and can drop it after the filter
 * You might load a column purely to use as a join key and drop it after the join
 * Etc.

So, rather than be forced to write a plan that is full of project nodes you can instead just emit at any point to drop a column once you are finished with it.

A corollary confusion is that the project node, by itself, cannot actually drop or reorder columns.  The project node, in Substrait, only adds columns to the end.  The only way to drop columns in a project node is with an emit.

This means that `SELECT expr1, expr2, expr3` typically gets represented as:

`project(expr1, expr2, expr3):emit=[len(input)+0, len(input)+1, len(input)+2]`

wjones127 commented on site/docs/tutorial/sql_to_substrait.md:
```diff
@@ -0,0 +1,1305 @@
+
+So first we can create the three extension uris:
+
+<!-- TODO: what is the URI format? -->
```
All the plans I have seen from ibis-substrait omits the URI, but that seems... wrong? Have we established the pattern for these URIs?

westonpace replied:
> Have we established the pattern for these URIs?

Not quite, but the time is ripe to fix this.  All produers I've seen (Isthmus & DuckDb) operate the same as Ibis.  Originally the contention was between:

 * Brittle URIs that are tied to a specific version of an extension function but the content is fixed (e.g. https://github.com/substrait-io/substrait/blob/v0.18.0/extensions/functions_arithmetic.yaml)
 * Stable URIs that are not tied to a specific version and the content may vary (e.g. https://github.com/substrait-io/substrait/blob/main/extensions/functions_arithmetic.yaml)

Based on the conversation in https://github.com/substrait-io/substrait/issues/274 I believe we were leaning towards the latter (stable URI, potentially changing content).

Note that, in Acero, we treat the URI `/` (or the empty string) as a special case where we fallback to name-only lookup and look for a function with the given name, regardless of what URI it maps to (and if there are multiple I think we pick the first we find).  This allows us to consume plans from producers today that omit the URI.

wjones127 commented on site/docs/tutorial/sql_to_substrait.md:
````diff
@@ -0,0 +1,1305 @@
+```
+
+Here, `names` is the names of all fields. In nested schemas, this includes the
+names of subfields in depth-first order. So for the `products` table, the 
+`details` struct field will be included as well as the two subfields (`manufacturer`
+and `year_created`) right after. And because it's depth first, these subfields
+appear before `product_name`. The full schema looks like:
````
Is this correct? There is no subfield for list or map types? (categories here is a list of strings)

westonpace replied:
I think so but I'm not certain.  I don't think a list column has names (even if it is a fixed size list) for its children.  In a map column the names are part of the data and not the field.

wjones127 commented on site/docs/tutorial/sql_to_substrait.md:
```diff
@@ -0,0 +1,1305 @@
+the `manufacturer` and `year_created` fields come right after their parent field,
+`details`.
+
+One complication is that whenever we reference by field index, those field
+indices are dependent on the context of the relation. The exact field indices
+can change from relation to relation. For example, a join relations field 
+indices are based on the combination of the left inputs fields and then the 
+right input fields. And, while we don't have one in our example, a Project relation
+adds new fields, which are visible to it's downstream relations. The diagram
+below shoes the mapping of field indices within each relation and how each of
+the field references show up in each relations properties.
```
Have I described the rules adequately / accurately? Is there anything important I've missed?

westonpace replied:
> those field indices are dependent on the context of the relation

> a join relations field indices

If I were to be pedantic I would say "the context of the expression".  A join relation does not have field indices.  A join relation has a "join expression" and the context of references in that expression is the output schema of the join relation.

An interesting case to consider (and I have a really old PR I need to get merged clarifying this :grimacing: https://github.com/substrait-io/substrait/pull/271/files) is the read relation.

A read relation has a filter expression and a projection (which isn't an expression exactly, but a masked expression).  A read relation effectively has _two_ schemas (maybe even more if we want to consider evolution but we won't go there :laughing: ).  There is the schema of the data on disk (the direct schema) and the output schema of the read relation (the direct schema after the projection mask has been applied to it).  So we have to be specific when talking about the filter expression which schema we are actually binding against.

wjones127 replied:
Thanks. It is complex, but I found that `substrait-validator` provides a really helpful UI for showing those schemas and so I've referenced that tool in the tutorial.

wjones127 reviewed (commented):
I have rewritten the tutorial to focus on translating a single example SQL query. It's still very rough, but before I go any further I wanted to get some technical feedback. I've noted a few areas where I'm not 100% sure I got it correct, so would like some feedback before I polish this up. @westonpace

vibhatha commented on site/docs/tutorial/sql_to_substrait.md:
```diff
@@ -0,0 +1,1305 @@
+
+Substrait is designed to communicate plans (mostly logical plans). Those plans
+contain types, schemas, expressions, extensions, and relations. We'll look at 
+them in that order, going from simplest to most complex until we can construct
+full plans.
```
nit: Suggestion

```suggestion
them in that order, going from basic to the most complex until we can construct
complete plans.
```

vibhatha commented on site/docs/tutorial/sql_to_substrait.md:
```diff
@@ -0,0 +1,1305 @@
+
+![](substrait_components.svg)
+
+This tutorial won't cover all the details of each piece, but will give you an
```
nit: maybe? 

```suggestion
This tutorial won't cover all the details of each piece, but it will give you an
```

vibhatha commented on site/docs/tutorial/sql_to_substrait.md:
```diff
@@ -0,0 +1,1305 @@
+![](plan_tree_versus_expression.svg)
+
+A relation that takes another relation as input will contain (or refer to) that
+relation. This is usually a field called `input`, but for complex cases that
+have multiple inputs such as joins may have different keys. In JSON, the rough 
```
Just want to know what you meant by this statement :)

wjones127 replied:
Yeah that is complex. I just meant that it's usually called `input`, but in the join case it's called `left` and `right`, so it's not always the same field for the inputs.

vibhatha commented on site/docs/tutorial/sql_to_substrait.md:
```diff
@@ -0,0 +1,1305 @@
+discuss later in [](#field-indices) where these come from.
+
+For `measures`, we'll need to define `sum(quantity * price) as sales`. Substrait
+is stricter about data types, and quantity is a integer while price is a
```
```suggestion
is stricter about data types, and quantity is an integer while price is a
```

vibhatha commented:
@wjones127 Thanks a lot for this tutorial, it is much needed component to docs. 👍

westonpace commented on site/docs/tutorial/sql_to_substrait.md:
```diff
@@ -0,0 +1,1305 @@
+them in that order, going from simplest to most complex until we can construct
+full plans.
+
+![](substrait_components.svg)
```
One minor nit about this image.  It shows four categories: types, relations, expressions, and functions.

I don't normally think of "functions" on the same level as the other three. Although, I'm also not sure exactly what "functions" is referring to, and maybe this is ok.

An expression can have a function invocation.  For example, a `ScalarFunction` message or a `WindowFunction` message.  But these are just a type of expression and so don't really deserve their own box.

On the other hand, a plan has a list of functions in the extension set (e.g. defines the URL/name/anchor).  If you're referring to this I suppose it's ok but in my mental picture I think of things like "read types" and "advanced extensions" at the same level of significance (e.g. not quite as significant as types/relations/expressions).

Anyways, I'm not sure anything here needs to change, it just seemed slightly off to me and I wanted to mention it.  If you're happy with it then we can stick with it as is.

wjones127 replied:
Yeah, this diagram is adapted from one of Jacques presentations. It's not really meant to accurately describe the relationships between each entity (except for the fact that plans are the top-level thing). It just gives an idea of what kinds of things are in plans.

I think it's fine for now, but if someone wants to design a better alternative, I have no qualms with replacing this.

wjones127 commented:
I have addressed the early PR feedback. I also validated the final query with substrait-validator (what an awesome tool! 🤩), and that let me know I was describing how field indices work incorrectly. I've revised the tutorial based on that feedback, and also added a "Next Steps" section pointing readers to that tool.

While there's a few TODOs left in there that would make this tutorial a bit better, I think this PR is ready for review.

wjones127 commented on site/docs/tutorial/sql_to_substrait.md:
````diff
@@ -0,0 +1,993 @@
+substrait-validator final_plan.json --out-file output.html
+```
+
+**Generate more Substrait plans from SQL using DuckDB's compiler**. See the 
+documentation on
+[the DuckDB Substrait extension](https://duckdb.org/docs/extensions/substrait).
````
I'd love to point readers to one of the tools that currently can be used to generate Substrait from SQL.

DuckDB looked easy to install, but not sure if they want that attention yet. I noticed the instructions on the website had some typos and the provided example errors when run in Python. From the github repo, they aren't accepting bug reports except from those who have pre-arranged agreements.

I didn't see a user-facing way to install Isthmus (just compilation instructions), but correct me if I'm wrong.

vibhatha replied:
Frankly I also think the existing docs could be improved further. I am not an expert in Graal based builds, but I also think it would be great to document the build, install and usage of isthmus with little more detail. I guess non-Java developers would also use this.

mbrobbel commented on site/docs/tutorial/final_plan.json:
```diff
@@ -0,0 +1,381 @@
+{
+  "version": { "minor": 18, "producer": "validator-test" },
```
The version field was renamed in 0.19.x via https://github.com/substrait-io/substrait/pull/362:
```suggestion
  "version": { "minorNumber": 19, "producer": "validator-test" },
```

westonpace commented on site/docs/tutorial/sql_to_substrait.md:
```diff
@@ -0,0 +1,995 @@
+The query asked the question: For products that aren't in the `"Computer"`
+category, how much has each product generated in sales?
+
+However, Substrait doesn't correspond to SQL as much as it does to logical plans.
```
```suggestion
However, Substrait doesn't correspond to SQL as much as it does to query plans.
```

Minor nit: Substrait could be used to represent both logical and physical plans.  This sentence _could_ potentially lead someone to thinking it is only for logical plans.

westonpace commented on site/docs/tutorial/sql_to_substrait.md:
````diff
@@ -0,0 +1,995 @@
+}
+```
+
+Nullability is an enum not a boolean, since Substrait allows 
+`NULLABILITY_UNSPECIFIED` as an option, in addition to `NULLABILITY_NULLABLE`
+(nullable) and `NULLABILITY_REQUIRED` (not nullable).
````
Eh...generally `_UNSPECIFIED` is an invalid value for an enum and will be rejected by all plans and consumers.  Pretty much every enum has an `_UNSPECIFIED` variant, even when it doesn't make sense (e.g. `JOIN_TYPE_UNSPECIFIED`).  This is done whenever an enum doesn't have a natural default value because otherwise there is no way in protobuf to distinguish between "use the default" and "this is missing".  So I'm not entirely sure why nullability is a boolean.  @jacques-n is `NULLABILITY_UNSPECIFIED` meant to be a legitimate thing?  I would think, if someone doesn't know, they would just pick `NULLABILITY_NULLABLE`.

westonpace commented on site/docs/tutorial/sql_to_substrait.md:
```diff
@@ -0,0 +1,995 @@
+https://github.com/substrait-io/substrait/tree/main/extensions. Additional
+extensions may be created elsewhere. `IS NULL` is defined as a `is_null` function
+in 
+[functions_comparison.yaml](https://github.com/substrait-io/substrait/blob/main/extensions/functions_comparison.yaml#:~:text=%2D-,name%3A%20%22is_null%22,-description%3A%20Whether)
+and `INDEX_IN` is defined as `index_in` function in
+[functions_set.yaml](https://github.com/substrait-io/substrait/blob/main/extensions/functions_set.yaml#:~:text=%2D-,name%3A%20%22index_in%22,-description%3A%20%3E).
```
```suggestion
[functions_comparison.yaml](https://github.com/substrait-io/substrait/blob/main/extensions/functions_comparison.yaml)
and `INDEX_IN` is defined as `index_in` function in
[functions_set.yaml](https://github.com/substrait-io/substrait/blob/main/extensions/functions_set.yaml).
```
Minor nit: I might be missing something but is the hash part of these URLs essential?

wjones127 replied:
That extra part is to highlight the relevant text. It means it will always scroll to the latest version of the function, regardless of where it is in the file.

westonpace commented on site/docs/tutorial/sql_to_substrait.md:
````diff
@@ -0,0 +1,995 @@
+```
+
+`functionReference` will be explained later in the plans section. For now,
+understand that it's an index into a list of function definitions and since this
````
Let's keep this unless you can come up with something better but this is technically not an "index into a list".  A functionReference of 7 could be the first item in the list or the 10th item.  It's more a key into a map.  However, I can't think of better wording without confusing a reader.

wjones127 replied:
I rephrased as an "ID corresponding to an entry in a list" which is hopefully clear and more accurate.

westonpace commented on site/docs/tutorial/sql_to_substrait.md:
````diff
@@ -0,0 +1,995 @@
+`outputType` defines the type the function outputs. We know this is a nullable
+`i64` type since that is what the function definition declares in the YAML file.
+
+`arguments` defines the arguments being passed into the function, which are all
+done positionally based on the function definition in the YAML file. The first 
+argument is an option that defines how `NaN` equality is handled, and since
+that's not relevant here we just pass `"NAN_IS_NOT_NAN"`. The next two options
+will be familiar as the literal and the field reference we constructed earlier.
+
+To create the final expression, we just need to wrap this in another scalar
+function expression for `IS NULL`.
+
+```json
+{
+  "scalarFunction": {
+    "functionReference": 2,
+    "outputType": {
+      "bool": {
+        "nullability": "NULLABILITY_REQUIRED"
+      }
+    },
+    "arguments": [
+      {
+        "value": {
+          "scalarFunction": {
+            "functionReference": 1,
+            "outputType": {
+              "i64": {
+                "nullability": "NULLABILITY_NULLABLE"
+              }
+            },
+            "arguments": [
+              {
+                "enum": {
+                  "specified": "NAN_IS_NOT_NAN"
````
Sorry, but this is out of date (my fault for the breaking change).  Options are no longer treated as arguments but rather their own thing.  So the `NAN_IS_NOT_NAN` will not be set in this way.

wjones127 replied:
Yup. Removed those references since they aren't required.

westonpace commented on site/docs/tutorial/sql_to_substrait.md:
```diff
@@ -0,0 +1,995 @@
+* For Aggregate relations, the output schema is the group by fields followed by
+  the measures.
+
+<!-- TODO: do we document these rules somewhere? -->
```
In the docs / spec this is referred to as the "direct output order"

![image](https://user-images.githubusercontent.com/1696093/206299475-17a02cc2-3f57-47e9-9e3d-bf027c62da6b.png)

wjones127 replied:
Also seems like "emit order" is used. I've added mention of those two phrases to the doc so readers will recognize them.

westonpace reviewed (changes requested):
Once again, this is a great document.  A few more thoughts but most are minor.

westonpace reviewed (approved):
Thanks for creating this!

westonpace reviewed (changes requested):
At the moment, you will need to patch up the commit messages to ensure they all confirm to semantic commits.  Some of your commit messages don't quite match.  Either squash the invalid ones into valid ones or reword the invalid ones.

wjones127 commented:
Thanks @westonpace. Just squashed all into one commit with what should be a conforming message.
